### PR TITLE
std: Add a new `env` module

### PR DESCRIPTION
**Pull request by alexcrichton**
This is an implementation of [RFC 578](https://github.com/rust-lang/rfcs/pull/578) which adds a new `std::env` module
to replace most of the functionality in the current `std::os` module. More
details can be found in the RFC itself, but as a summary the following methods
have all been deprecated:
- `os::args_as_bytes`   => `env::args`
- `os::args`            => `env::args`
- `os::consts`          => `env::consts`
- `os::dll_filename`    => no replacement, use `env::consts` directly
- `os::page_size`       => `env::page_size`
- `os::make_absolute`   => use `env::current_dir` + `join` instead
- `os::getcwd`          => `env::current_dir`
- `os::change_dir`      => `env::set_current_dir`
- `os::homedir`         => `env::home_dir`
- `os::tmpdir`          => `env::temp_dir`
- `os::join_paths`      => `env::join_paths`
- `os::split_paths`     => `env::split_paths`
- `os::self_exe_name`   => `env::current_exe`
- `os::self_exe_path`   => use `env::current_exe` + `pop`
- `os::set_exit_status` => `env::set_exit_status`
- `os::get_exit_status` => `env::get_exit_status`
- `os::env`             => `env::vars`
- `os::env_as_bytes`    => `env::vars`
- `os::getenv`          => `env::var` or `env::var_string`
- `os::getenv_as_bytes` => `env::var`
- `os::setenv`          => `env::set_var`
- `os::unsetenv`        => `env::remove_var`

Many function signatures have also been tweaked for various purposes, but the
main changes were:
- `Vec`-returning APIs now all return iterators instead
- All APIs are now centered around `OsString` instead of `Vec<u8>` or `String`.
  There is currently on convenience API, `env::var_string`, which can be used to
  get the value of an environment variable as a unicode `String`.

All old APIs are `#[deprecated]` in-place and will remain for some time to allow
for migrations. The semantics of the APIs have been tweaked slightly with regard
to dealing with invalid unicode (panic instead of replacement).

The new `std::env` module is all contained within the `env` feature, so crates
must add the following to access the new APIs:

```
#![feature(env)]
```

[breaking-change]

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

note that https://github.com/rust-lang/rfcs/pull/578 has not yet been merged and this should hold off on approval until that time.

File: src/libstd/env.rs [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+/// ```rust
+/// use std::env;
+///
+/// // We assume that we are in a valid directory.
+/// let p = env::current_dir().unwrap();
+/// println!("The current directory is {}", p.display());
+/// ```
+pub fn current_dir() -> IoResult<Path> {
+    os_imp::getcwd()
+
```
Comment (nagisa) [inline src/libstd/env.rs]: Extra newline.

File: src/libstd/env.rs [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+            }
+        }
+    }
+}
+
+impl Error for VarError {
+    fn description(&self) -> &str {
+        match *self {
+            VarError::NotPresent => "environment variable not found",
+            VarError::NotUnicode(..) => "environment variable not valid unicode",
```
Comment (nagisa) [inline src/libstd/env.rs]: > environment variable was not not valid unicode

to match what Display impl outputs.

File: src/libstd/env.rs [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+        env_path("TMPDIR").unwrap_or(if cfg!(target_os = "android") {
+            Path::new("/data/local/tmp")
+        } else {
+            Path::new("/tmp")
+        })
+    } else {
+        env_path("TMP").or(env_path("TEMP"))
+                       .or(env_path("USERPROFILE"))
+                       .or(env_path("WINDIR"))
+                       .unwrap_or(Path::new("C:\\Windows"))
```
Comment (nagisa) [inline src/libstd/env.rs]: This behaviour might be surprising, but it matches [GetTempPath](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364992%28v=vs.85%29.aspx) Winapi function, but I’m not sure about assuming [GetWindowsDirectory](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724454%28v=vs.85%29.aspx) will always be `C:\Windows`.

**Comment by nagisa** [comment]
LGTM

File: src/libstd/env.rs [hunk 4]
```diff
@@ -0,0 +1,858 @@
...
+    Vars { inner: os_imp::env() }
+}
+
+impl Iterator for Vars {
+    type Item = (OsString, OsString);
+    fn next(&mut self) -> Option<(OsString, OsString)> { self.inner.next() }
+    fn size_hint(&self) -> (usize, Option<usize>) { self.inner.size_hint() }
+}
+
+/// Fetches the environment variable `n` from the current process, returning
```
Comment (aturon) [inline src/libstd/env.rs]: `n` should be `k`

File: src/libstd/env.rs [hunk 5]
```diff
@@ -0,0 +1,858 @@
...
+///     Some(val) => println!("{}: {:?}", key, val),
+///     None => println!("{} is not defined in the environment.", key)
+/// }
+/// ```
+pub fn var<K: ?Sized>(key: &K) -> Option<OsString> where K: AsOsStr {
+    let _g = ENV_LOCK.lock();
+    os_imp::getenv(key.as_os_str())
+}
+
+/// Fetches the environment variable `n` from the current process.
```
Comment (aturon) [inline src/libstd/env.rs]: `n` should be `key`

File: src/libstd/env.rs [hunk 6]
```diff
@@ -0,0 +1,858 @@
...
+impl Error for VarError {
+    fn description(&self) -> &str {
+        match *self {
+            VarError::NotPresent => "environment variable not found",
+            VarError::NotUnicode(..) => "environment variable was not valid unicode",
+        }
+    }
+}
+
+/// Sets the environment variable `n` to the value `v` for the currently running
```
Comment (aturon) [inline src/libstd/env.rs]: `n` should be `k`

File: src/libstd/env.rs [hunk 7]
```diff
@@ -0,0 +1,858 @@
...
+/// Optionally returns the filesystem path to the current executable which is
+/// running but with the executable name.
+///
+/// # Examples
+///
+/// ```rust
+/// use std::env;
+///
+/// match env::current_exe() {
+///     Some(exe_path) => println!("Path of this executable is: {}",
```
Comment (aturon) [inline src/libstd/env.rs]: This example is assuming `Option`.

Might be helpful to say when this can fail.

File: src/libstd/sys/unix/os.rs [hunk 8]
```diff
@@ -11,93 +11,62 @@
 //! Implementation of `std::os` functionality for unix systems
 
 use prelude::v1::*;
+use os::unix::*;
```
Comment (aturon) [inline src/libstd/sys/unix/os.rs]: :D

**Comment by aturon** [comment]
This basically LGTM, modulo tiny nits. (FYI: I didn't scrutinize the `sys` changes super carefully)

File: src/libstd/sys/windows/backtrace.rs [hunk 9]
```diff
@@ -7,29 +7,31 @@
...
+//! If you google around, you'll find a fair bit of references to built-in
+//! functions to get backtraces on windows. It turns out that most of these are
+//! in an external library called dbghelp. I was unable to find this library
+//! via `-ldbghelp`, but it is apparently normal to do the `dlopen` equivalent
+//! of it.
+//!
+//! You'll also find that there's a function called CaptureStackBackTrace
+//! mentioned frequently (which is also easy to use), but sadly I didn't have a
+//! copy of that function in my mingw install (maybe it was broken?). Instead,
+//! this takes the route of using StackWalk64 in order to walk the stack.
```
Comment (retep998) [inline src/libstd/sys/windows/backtrace.rs]: Did you know the function is actually `RtlCaptureStackBackTrace` and that `CaptureStackBackTrace` is merely `#define`d to that? I'll have to try writing a PR to see if I can get stacktraces working using this.

File: src/libstd/sys/windows/fs.rs [hunk 10]
```diff
@@ -368,18 +364,18 @@ pub fn readlink(p: &Path) -> IoResult<Path> {
...
         GetFinalPathNameByHandleW(handle,
                                   buf as *const u16,
                                   sz - 1,
                                   libc::VOLUME_NAME_DOS)
-    });
+    }));
+    let ret = String::from_utf16(&*ret).unwrap();
     let ret = match ret {
-        Some(ref s) if s.starts_with(r"\\?\") => { // "
+        ref s if s.starts_with(r"\\?\") => { // "
```
Comment (retep998) [inline src/libstd/sys/windows/fs.rs]: Why are we getting rid of the verbatim prefix?

File: src/libstd/sys/windows/backtrace.rs (hunk 9 above)
Comment (alexcrichton) [inline src/libstd/sys/windows/backtrace.rs]: I'm... not actually sure! Most of this fallout was from removing all the `#[allow]` attributes in the `windows/mod.rs` file, I didn't do many substantial changes to them. That being said, I remember when I initially wrote this I basically took the first API that looked reasonable, I wouldn't say I know what I was doing at the time! If functions like this are normally `#define`'d then that's probably what happened (and we may be able to move to it)

File: src/libstd/sys/windows/fs.rs (hunk 10 above)
Comment (alexcrichton) [inline src/libstd/sys/windows/fs.rs]: I'm... not sure. This seems to have been introduced in 40ab198356bbec05b308b2e61ebc6fe9e23ade9d, but I have no idea why!

When the `std::fs` RFC lands I'll try to be sure to see if we can remove this logic.

File: src/libstd/env.rs (hunk 3 above)
Comment (alexcrichton) [inline src/libstd/env.rs]: Given that there's a system API to call here, sounds like we should be calling that! I will change the implementation to call this.

File: src/libstd/env.rs [hunk 11]
```diff
@@ -0,0 +1,853 @@
...
+/// ```rust
+/// use std::env;
+///
+/// match env::home_dir() {
+///     Some(ref p) => println!("{}", p.display()),
+///     None => println!("Impossible to get your home dir!")
+/// }
+/// ```
+pub fn home_dir() -> Option<Path> {
+    env_path("HOME").or(if cfg!(windows) {env_path("USERPROFILE")} else {None})
```
Comment (nagisa) [inline src/libstd/env.rs]: So, if we’re going the path of using WinAPI where possible [GetUserProfileDirectory](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762280%28v=vs.85%29.aspx) is probably what should be used here.

Comment (retep998) [inline src/libstd/env.rs]: `GetUserProfileDirectoryW` is actually part of `userenv.lib`, which we don't bundle with Rust on Windows at the moment, although we certainly could and the mingw64 library `libuserenv.a` has that function.

Comment (nagisa) [inline src/libstd/env.rs]: Apparently on POSIX the story doesn’t end at `$HOME` either. If `$HOME` is not set `passwd` should be queried for the home directory.
```
struct passwd pwd, *result;
char buf[1000];
if (getpwuid_r(getuid(), &pwd, buf, 999, &result) >= 0) 
    printf("%s", result->pw_dir);
```

Comment (retep998) [inline src/libstd/env.rs]: The really nice thing about `GetUserProfileDirectoryW` is that it doesn't use environment variables but rather queries the internal user data to get their profile directory, and can even get profile directories for arbitrary users.

Comment (alexcrichton) [inline src/libstd/env.rs]: Thanks for the pointers! I've updated both the unix and windows implementations.

**Comment by John-Nagle** [comment]
This is great! With this, to get the command line arguments as a vector of strings, you simply write

    let args: Vec<String> = env::args().map(|s| s.into_string().unwrap()).collect();

Now, new programmers learning Rust will have to understand closures, iterators, functional programming, and UTF8 vs OS strings just to get past "Hello World". That will keep out the Code Academy riff-raff!

Put in a convenience function in env::args for this. 

For a sense of the pain caused by this, see this Stack Overflow question:

http://stackoverflow.com/questions/15619320/how-to-access-command-line-parameters-in-rust/28356205#28356205

**Comment by nagisa** [comment]
@John-Nagle your example panics if any of the variables contain non-unicode data.

**Comment by John-Nagle** [comment]
I know. "pub fn into_string_lossy(Self) -> String;" for OsString doesn't seem to have been implemented yet. 

If you have non-Unicode data from OsString conversion of command line args, something is badly broken.

**Comment by nagisa** [comment]
> If you have non-Unicode data from OsString conversion of command line args, something is badly broken.

Are you implying any system which uses non utf-8 locale is broken?

**Comment by aturon** [comment]
@John-Nagle 

I'm sympathetic to your original comment; the prominence of `OsString` was debated as part of this API, and what landed in this PR was a minimalistic, initial cut.

An alternative API that pushes stronger for Unicode-by-default might be:

```rust
pub fn args() -> Args // Iterator<Item = String>
pub fn args_os() -> ArgsOs // Iterator<Item = OsString>

pub fn var<K: ?Sized>(key: &K) -> Result<String, VarError> where K: AsOsStr
pub fn var_os<K: ?Sized>(key: &K) -> Option<OsString> where K: AsOsStr

pub fn vars() -> Vars // Iterator<Item = (String, String)>
pub fn vars_os() -> VarsOs // Iterator<Item = (OsString, OsString)>
```

where the non-OS versions will panic on non-Unicode data, just as your example does. We're likely to grow such variants regardless; the main question is what the un-decorated names like `args` should map to.

While I'm also sympathetic to worries about how much newcomers have to learn at once, I don't particularly buy the point about `Vec` versus iterators. For one, iterators are an extremely common concept found in many (these days, most) languages. It's a concept you need to know to understand `for` in Rust. For another, if you're thinking about very simple scripts that take e.g. a single argument, then `args().nth(1)` will do the trick (and give you an `Option` value that you can use to display an error for example).

> I know. "pub fn into_string_lossy(Self) -> String;" for OsString doesn't seem to have been implemented yet.

The `to_string_lossy` method is probably what you want (http://static.rust-lang.org/doc/master/std/ffi/struct.OsStr.html#method.to_string_lossy), but discoverability is a bit poor right now because rustdoc doesn't yet hint about `Deref`.

**Comment by John-Nagle** [comment]
*Are you implying any system which uses non utf-8 locale is broken?*
No, that's handled. OsString is an opaque type. See
https://github.com/rust-lang/rfcs/blob/master/text/0517-io-os-reform.md#the-design-os_str
especially where it says "but CRUCIALLY NOT as_bytes". If OsString values are locale sensitive, the OS support package for that platform must deal with it invisibly to the Rust program.

This whole area seems way overdesigned. The old approach was much simpler, and the use case for all this complexity seems to be corrupted UCS-2 on Windows. Anything Windows can represent as UCS-2, even with surrogate pairs, can be represented as a UTF-8 Rust string. Some conversions in the other direction won't work, and they should raise exceptions. Yes, this may mean that you can't put emoji into a Windows file name.

**Comment by aturon** [comment]
> This whole area seems way overdesigned. The old approach was much simpler, and the use case for all this complexity seems to be corrupted UCS-2 on Windows.

The other side of `OsString`, on Unix, is dealing with byte sequences that are not UTF-8. Since paths, arguments, environment variables, etc are just C strings (and in many systems are not required to follow any particular encoding), we need a way to represent the full space of values.

Note that "the old approach" already had things like `env_as_bytes` to tackle the above. This has just been packaged into a more robust, cross-platform `OsString`.

That said, as I wrote above, I think we probably want to consider making the "default" variants work with `String` and have a separate set of more "advanced" forms that robust, cross-platform code would use.

**Comment by nagisa** [comment]
@John-Nagle yes, this is handled by `OsString`, but conversion of the data inside `OsString` into `String` will fail with panic in many cases.

Which is exactly why I argue against any methods converting to `String` implicitly.

**Comment by aturon** [comment]
@John-Nagle 

> Anything Windows can represent as UCS-2, even with surrogate pairs, can be represented as a UTF-8 Rust string. Some conversions in the other direction won't work, and they should raise exceptions. 

This appears to be backwards. UCS-2 is a superset of UTF-16. That means that e.g. Rust strings can be converted to UCS-2 without loss, but not vice versa.

Regardless, the motivation for `OsString` is to provide some means of faithfully representing the platform's "native" string type (usually UCS-2 on Windows, byte sequences on Unixen) without loss and in a way that supports cross-platform programming.

**Comment by John-Nagle** [comment]
> That said, as I wrote above, I think we probably want to consider making the "default" variants work with String and have a separate set of more "advanced" forms that robust, cross-platform code would use.

That seems to be the most practical approach. New Rust programmers should not have to comprehend all the topics covered in this discussion just to access the command line arguments.

**Pull request merged**